typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1655 @@
+
+#### Tagless final eDSL
+
+Another way to implement an eDSL is [tagless final](https://serokell.io/blog/introduction-tagless-final).
+We can move from data type to type classes with the same logic.
+Here the type class will describe possible actions, and by using instances, we will be able to write interpreters.
+
+Let's start by defining our type class.
+To do this, we need the already-mentioned `URIS`(which is a union of all 1-arity defined in `URItoKind`) and the already-mentioned `Kind`.
+
+```typescript
+import { Kind, URIS } from "fp-ts/HKT";
+
+type ArithExpr<Expr extends URIS> = {
+  num: (v: number) => Kind<Expr, number>;
+  plus: (l: Kind<Expr, number>, r: Kind<Expr, number>) => Kind<Expr, number>;
+  gt: (l: Kind<Expr, number>, r: Kind<Expr, number>) => Kind<Expr, boolean>;
+  and: (l: Kind<Expr, boolean>, r: Kind<Expr, boolean>) => Kind<Expr, boolean>;
+};
+```
+
+Here are the types for which we will create `ArithExpr` instances:
+
+* `Interpreter<A>`, which defines how the algebra should behave.
+* `ToS<A>`, which will create a string representation of the expression.
+
+We need to somehow add them to the `fp-ts` `URItoKind` interface to be able to use them in `ArithExpr` since it takes `Expr extends URIS`, where `URIS` is just `keyof URItoKind<unknown>`.
+
+Here we need the TypeScript's module augmentation feature, which allows us to patch existing objects by importing and then updating them.
+So, we simply add `Interpreter` and `ToS` to `URItoKind`.
+
+```typescript
+type Interpreter<A> = {
+  interpret: A;
+};
+
+type ToS<A> = {
+  toString: string;
+};
+
+declare module "fp-ts/lib/HKT" {
+  interface URItoKind<A> {
+    readonly Interpreter: Interpreter<A>;
+    readonly ToS: ToS<A>;
+  }
+}
+```
+
+After all the preparations, we are ready to create instances.
+
+```typescript
+const arithInterpreter: ArithExpr<"Interpreter"> = {
+  num: (v: number) => {
+    return { interpret: v };
+  },
+  plus: (l: Interpreter<number>, r: Interpreter<number>) => {
+    return { interpret: l.interpret + r.interpret };
+  },
+  gt: (l: Interpreter<number>, r: Interpreter<number>) => {
+    return { interpret: l.interpret > r.interpret };
+  },
+  and: (l: Interpreter<boolean>, r: Interpreter<boolean>) => {
+    return { interpret: l.interpret && r.interpret };
+  },
+};
+
+const arithToS: ArithExpr<"ToS"> = {
+  num: (v: number) => {
+    return { toString: v.toString() };
+  },
+  plus: (l: ToS<number>, r: ToS<number>) => {
+    return { toString: `(${l.toString} + ${r.toString})` };
+  },
+  gt: (l: ToS<number>, r: ToS<number>) => {
+    return { toString: `(${l.toString} > ${r.toString})` };
+  },
+  and: (l: ToS<boolean>, r: ToS<boolean>) => {
+    return { toString: `(${l.toString} && ${r.toString})` };
+  },
+};
```

Comment:
My wild guess is that `"ToS"` here is a marker or has the same role as
```
instance IsExpr ToS where
```
in haskell.

Reply:
If I understand you right the answer is yes \
I don't really like how it looks, but I wanted to make it looks like in haskell and didn't want to invent more complex examples as it is already complex enough. So, I thought a lot about it and as it is kind of advanced part I think we can leave it like this